XAMLator.Server.Shared/MonoEvaluator.cs

```diff
@@ -6,110 +6,168 @@
-		Printer printer;
-
-		public Task<bool> EvaluateExpression(string expression, string code, EvalResult result)
-		{
-			EnsureConfigured();
-			try
-			{
-				object retResult;
-				bool hasResult;
-
-				printer.Reset();
-				if (!String.IsNullOrEmpty(code))
-				{
-					var ret = eval.Evaluate(code, out retResult, out hasResult);
-				}
-				result.Result = eval.Evaluate(expression);
-				return Task.FromResult(true);
-			}
-			catch (Exception ex)
-			{
-				Log.Error($"Error creating a new instance of {expression}");
-				if (printer.Messages.Count != 0)
-				{
-					result.Messages = printer.Messages.ToArray();
-				}
-				else
-				{
-					result.Messages = new EvalMessage[] { new EvalMessage("error", ex.ToString()) };
-				}
-				if (!result.HasResult && result.Messages.Length == 0)
-				{
-					result.Messages = new EvalMessage[] { new EvalMessage("error", "Internal Error") };
-				}
-				eval = null;
-			}
-			return Task.FromResult(false);
-		}
-
-		void EnsureConfigured()
-		{
-			if (eval != null)
-			{
-				return;
-			}
-
-			var settings = new CompilerSettings();
-			printer = new Printer();
-			var context = new CompilerContext(settings, printer);
-			eval = new Mono.CSharp.Evaluator(context);
-			AppDomain.CurrentDomain.AssemblyLoad += (_, e) =>
-			{
-				LoadAssembly(e.LoadedAssembly);
-			};
-			foreach (var a in AppDomain.CurrentDomain.GetAssemblies())
-			{
-				LoadAssembly(a);
-
-			}
-		}
-
-		void LoadAssembly(Assembly assembly)
-		{
-			var name = assembly.GetName().Name;
-			if (name == "mscorlib" || name == "System" || name == "System.Core")
-				return;
-			eval?.ReferenceAssembly(assembly);
-		}
-	}
-
-	class Printer : ReportPrinter
-	{
-		public readonly List<EvalMessage> Messages = new List<EvalMessage>();
-
-		public new void Reset()
-		{
-			Messages.Clear();
-			base.Reset();
-		}
-
-		public override void Print(AbstractMessage msg, bool showFullPath)
-		{
-			if (msg.MessageType != "error")
-			{
-				return;
-			}
-			AddMessage(msg.MessageType, msg.Text, msg.Location.Row, msg.Location.Column);
-		}
-
-		public void AddError(Exception ex)
-		{
-			AddMessage("error", ex.ToString(), 0, 0);
-		}
-
-		void AddMessage(string messageType, string text, int line, int column)
-		{
-			var m = new EvalMessage(messageType, text, line, column);
-			Messages.Add(m);
-			if (m.MessageType == "error")
-			{
-				Log.Error(m.Text);
-			}
-		}
-	}
+    public class Evaluator : IEvaluator
+    {
+        Mono.CSharp.Evaluator eval;
+        Printer printer;
+        bool canEvalutate = false;
+
+        public Task<bool> EvaluateExpression(string expression, string code, EvalResult result)
+        {
+            EnsureConfigured();
+            try
+            {
+                //on real devices we cannot do eval.Evaluate! https://github.com/mono/mono/issues/6616
+                if (!canEvalutate)
+                {
+                    //TODO: do we really need expression? why not just pass the type to instantiate
```

Comment:
We need to pass the expression string because the type might not have a default constructor.
If evaluation is not possible it shouldn't even try to evaluate anything.

To support this scenario I would do to the following:
- Add a new field to EvalRequest named OriginalTypeName that includes the full name of the Type 
- In VM.cs, if the platform does not support evaluation, create a new instance by type using the OriginalTypeName